### PR TITLE
Add Awakener

### PR DESCRIPTION
**Pull request by Thomasdezeeuw**
Awakener allows cross-thread waking of Poll.

This is aport of Gaea's Awakener.

Updates #968.

### CODE REVIEW

**Comment by Thomasdezeeuw** [comment]
@asomers Can you check kqueue's side of this?

**Comment by Thomasdezeeuw** [comment]
I forgot to run Rustfmt, will do after some initial feedback.

**Comment by Thomasdezeeuw** [comment]
Builds are failing because of fmt and various warnings not related to this change. I'll fix that in another pr I'm currently working on.

**Comment by Thomasdezeeuw** [comment]
Solaris is Linux like right? Does that mean it has eventfd?

**Comment by Thomasdezeeuw** [comment]
This also need the suppression of warnings in #970 for the tests to pass.

**Comment by Thomasdezeeuw** [comment]
I've rebased this on #970, so hopefully the CI will be green.

**Comment by Thomasdezeeuw** [comment]
I've used a unix pipe for Solaris. Now all that is left is Windows. It seems to get the wrong token: https://dev.azure.com/tokio-rs/Tokio/_build/results?buildId=867&view=logs&jobId=429a383f-e486-57ac-3982-9ac65641a32a.

File: src/awakener.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+///     // Now we'll wake the queue on the other thread.
+///     awakener1.wake().expect("unable to wake");
+/// });
+///
+/// // On our current thread we'll poll for events, without a timeout.
+/// poll.poll(&mut events, None)?;
+///
+/// // After about 500 milliseconds we should we awoken by the other thread we
+/// // started, getting a single event.
+/// for event in &events {
```
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/awakener.rs]: You're right, but `Events` doesn't provide a length method. I'll add `assert!(!events.is_empty())`.

File: src/sys/unix/dlsym.rs [hunk 2]
```diff
@@ -16,13 +16,15 @@ macro_rules! dlsym {
     )
 }
 
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/sys/unix/dlsym.rs]: On macOS this gave errors, but it's really unrelated to adding the Awakener.

File: test/test_awakener.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    // Waking on the same thread.
+    awakener.wake().expect("unable to wake");
+    expect_events(
+        &mut poll,
+        &mut events,
+        1,
+        vec![Event::new(Ready::readable(), token)],
+    );
+
+    // Multiple wakes between polls.
```
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply test/test_awakener.rs]: Will do.

**Comment by Thomasdezeeuw** [comment]
> It strikes me that in the spirit of slimming mio down, Awakener could be moved to a separate crate. I'm not going to insist upon it, though.

@asomers You're right but it's fairly common to require this behaviour. And it does need add to the Selector internal on kqueue. So even if this would go into a separate crate it would be rather tightly coupled.

I'll address the comments and rebase on master.

File: src/sys/unix/dlsym.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (asomers) [reply src/sys/unix/dlsym.rs]: It would be better to remove the symbol on platforms where its unused.  That's preferable to `#[allow(dead_code)]` IMHO.

**Comment by asomers** [comment]
> > It strikes me that in the spirit of slimming mio down, Awakener could be moved to a separate crate. I'm not going to insist upon it, though.
> 
> @asomers You're right but it's fairly common to require this behaviour. And it does need add to the Selector internal on kqueue. So even if this would go into a separate crate it would be rather tightly coupled.
> 
> I'll address the comments and rebase on master.

Makes sense.

**Comment by Thomasdezeeuw** [comment]
I've force pushed the changes, the last commit addresses @asomers comments.

File: src/sys/unix/dlsym.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/sys/unix/dlsym.rs]: This change is also in https://github.com/tokio-rs/mio/commit/434405a66c66c1e2b15a07388dd0ed4edcc7e3c8, which is already committed to master. But I do agree it's not a great solution.

**Comment by carllerche** [comment]
@asomers What to include in `mio` proper is a bit of a balance. Should TCP / UDP be included? It is common enough it seems like it. What about UDS? It is part of stdlib, i think we should consider it.

**Comment by carllerche** [comment]
"Slimming down" is equally about maintenance as well as code in the hot path.

**Comment by Thomasdezeeuw** [comment]
@asomers Sorry about the force push.

Windows doesn't work (see https://dev.azure.com/tokio-rs/Tokio/_build/results?buildId=894&view=logs&jobId=429a383f-e486-57ac-3982-9ac65641a32a), it returns the incorrect token. It seems to return the max value of usize.

**Comment by Thomasdezeeuw** [comment]
Who has a Windows machine that can help out with getting the Windows side working?

File: src/awakener.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+use crate::{poll, sys, Registry, Token};
+
+/// Awakener allows cross-thread waking of [`Poll`].
+///
+/// When created it will cause events with [`Ready::readable()`] and the provided
+/// `token` if [`wake`] is called, possibly from another thread.
+///
+/// # Notes
+///
+/// The `Awakener` needs to be kept alive as long as wake up notifications are
```
Comment (carllerche) [inline src/awakener.rs]: I would reword this as stating that wake up events are only guaranteed to be delivered while the `Awakener` value is alive.

Comment (carllerche) [reply src/awakener.rs]: The implementation detail isn't relevant for the public API docs imo

**Comment by Thomasdezeeuw** [comment]
@carllerche suggested the name `Waker` instead of `Awakener`, any thoughts @asomers?

File: src/awakener.rs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+/// required. This is due to an implementation detail where if all copies of the
+/// `Awakener` are dropped it will also drop all wake up notifications from the
+/// system queue, including wake up notifications that have been added before
+/// the `Awakener` that was dropped, resulting the [`Poll`] not being woken up.
+///
+/// Only a single `Awakener` should active per [`Poll`], if multiple threads
+/// need access to the `Awakener` it can be shared via for example an `Arc`.
+/// What happens if multiple `Awakener`s are registered with the same `Poll` is
```
Comment (carllerche) [inline src/awakener.rs]: Why is it undefined? If they are registered with different tokens, it should be OK?

Comment (Thomasdezeeuw) [reply src/awakener.rs]: I put that in to allow some room for future implementations that weren't ok with multiple `Awakener`s, but currently it's fine on all platforms I think

**Comment by asomers** [comment]
> @carllerche suggested the name `Waker` instead of `Awakener`, any thoughts @asomers?

Either name is fine by me.

File: src/awakener.rs (hunk 4 above)
Comment (Thomasdezeeuw) [reply src/awakener.rs]: Keeping the mentioning of it being an implementation detail (second line)?

**Comment by Thomasdezeeuw** [comment]
Ok then I'll rename it to `Waker`, but that means I'll also have to rename some files to I'll leave if we're all on agreement on the other stuff.

File: src/awakener.rs (hunk 5 above)
Comment (Thomasdezeeuw) [reply src/awakener.rs]: I can remove it if you want.

Comment (carllerche) [reply src/awakener.rs]: I'm just wondering about it. I was thinking, since tokens are exposed, you could have multiple wakers that have different significance.

File: src/sys/unix/kqueue.rs [hunk 6]
```diff
@@ -238,6 +238,78 @@ impl Selector {
...
+        if (kevent.flags & libc::EV_ERROR) != 0 && kevent.data != 0 {
+            Err(io::Error::from_raw_os_error(kevent.data as i32))
+        } else {
+            Ok(())
+        }
+    }
+
+    // Used by `Awakener`.
+    #[cfg(any(target_os = "freebsd", target_os = "ios", target_os = "macos"))]
+    pub fn try_clone(&self) -> io::Result<Selector> {
```
Comment (carllerche) [inline src/sys/unix/kqueue.rs]: Since this is kqueue specific, what if we don't return a `Selector`? That would prevent needing incrementing `NEXT_ID`. We could just return the RawFd, or some other private type.

File: src/sys/windows/awakener.rs [hunk 7]
```diff
@@ -1,73 +1,39 @@
...
-        events: Interests,
-        opts: PollOpt,
-    ) -> io::Result<()> {
-        self.register(registry, token, events, opts)
-    }
-
-    fn deregister(&self, _registry: &Registry) -> io::Result<()> {
-        *self.inner.lock().unwrap() = None;
+        let inner = self.inner.lock().unwrap();
+        let status = CompletionStatus::new(0, usize::from(inner.token), 0 as *mut _);
```
Comment (carllerche) [inline src/sys/windows/awakener.rs]: The windows implementation does not work as one would expect. Receiving completion status events in `select` does not result in a mio event. Instead, in `select`, the callback is called which then results in an event getting posted via the custom readiness queue.

The easiest way to hack in awakeners would probably be to use a hard coded awakener token (there already is one [here](https://github.com/tokio-rs/mio/blob/master/src/sys/windows/selector.rs#L87)). Then, use some other field to pass in the token... the best bet is probably to use the `lpOverlapped` pointer (set to 0 here).

Then, in `select`, when an awakener event is received, push an appropriate event into the events list.

Hopefully we can replace the windows impl soon...

File: src/sys/unix/kqueue.rs (hunk 6 above)
Comment (Thomasdezeeuw) [reply src/sys/unix/kqueue.rs]: We can also just use `0`/max as id since it won't be used in the `Awakener` anyway.

File: src/sys/windows/awakener.rs (hunk 7 above)
Comment (Thomasdezeeuw) [reply src/sys/windows/awakener.rs]: Can't we use the `Vec` in `Events`? We detect `overlapped() == 0` in `select` already.

Comment (Thomasdezeeuw) [reply src/sys/windows/awakener.rs]: I've just pushed https://github.com/tokio-rs/mio/pull/969/commits/37a6ba282aad1fe3ace8d26325f6387eed45f3d6, see if it works.

**Comment by Thomasdezeeuw** [comment]
Ok that worked! It just broke other tests on Windows :)

File: src/sys/windows/awakener.rs (hunk 7 above)
Comment (Thomasdezeeuw) [reply src/sys/windows/awakener.rs]: This works, finally.

File: src/awakener.rs (hunk 5 above)
Comment (Thomasdezeeuw) [reply src/awakener.rs]: Want me to remove it?

**Comment by Thomasdezeeuw** [comment]
CI is finally green.

File: src/sys/unix/kqueue.rs (hunk 6 above)
Comment (Thomasdezeeuw) [reply src/sys/unix/kqueue.rs]: Reusing the id seems to work fine, so I'll do that.

File: src/awakener.rs (hunk 5 above)
Comment (Thomasdezeeuw) [reply src/awakener.rs]: I've tried adding a test for this and it doesn't work on kqueue platforms currently.

Comment (Thomasdezeeuw) [reply src/awakener.rs]: I could try and fix that.

**Comment by Thomasdezeeuw** [comment]
Don't want me to rename it to `Waker`?

**Comment by carllerche** [comment]
@Thomasdezeeuw follow up PR? I mostly wanted to get it in to fix the merge conflicts

**Comment by Thomasdezeeuw** [comment]
Sure

**Comment by Thomasdezeeuw** [comment]
Pr for renaming: https://github.com/tokio-rs/mio/pull/976.
